indexer-app/src/cmr/indexer/data/concepts/keyword_util.clj

```diff
@@ -20,5 +20,152 @@
+            (mapcat names->keywords instruments))))
+
+(defn related-url->keywords
+  "Converts a compound field into a vector of terms for keyword searches."
+  [data]
+  (let [{description :Description
+         subtype :Subtype
+         type :Type
+         url :URL
+         url-content-type :URLContentType} data]
+    [description
+     subtype
+     type
+     url
+     url-content-type]))
+
+(defn science-keyword->keywords
+  "Converts a science keyword into a vector of terms for keyword searches."
+  [science-keyword]
+  (let [{category :Category
+         detailed-variable :DetailedVariable
+         term :Term
+         topic :Topic
+         variable-level-1 :VariableLevel1
+         variable-level-2 :VariableLevel2
+         variable-level-3 :VariableLevel3} science-keyword]
+    [category
+     detailed-variable
+     term
+     topic
+     variable-level-1
+     variable-level-2
+     variable-level-3]))
+
+(defn service-keyword->keywords
+  "Converts a service keyword into a vector of terms for keyword searches."
+  [service-keyword]
+  (let [{service-category :ServiceCategory
+         service-specific-term :ServiceSpecificTerm
+         service-term :ServiceTerm
+         service-topic :ServiceTopic} service-keyword]
+    [service-category
+     service-specific-term
+     service-term
+     service-topic]))
+
+(defn service-organization->keywords
+  "Converts a service keyword into a vector of terms for keyword searches."
+  [service-organization]
+  (let [{roles :Roles
+         service-contact-persons :ContactPersons} service-organization]
+    (concat (names->keywords service-organization)
+            (mapcat contact-person->keywords service-contact-persons)
+            roles)))
+
+(def fields->fn-mapper
+  "A data structure that maps UMM field names to functions that extract keyword
+  data for those fields. Intended to be used instead of `case` statements for
+  dispatching based upon field name.
+
+  For example, to iterate over all the science keywords in a concept and return
+  textual data that will be indexd (i.e., from sub-fields):
+
+    (map (:ScienceKeywords fields->fn-mapper) parsed-concept))"
+  {;; Simple single-valued data
+   :variable-name :variable-name
+   :measurement :measurement
+   :LongName :LongName
+   :Name :Name
+   :Version :Version
+   ;; Nested single-value data
+   :RelatedURL #(related-url->keywords (:RelatedURL %))
+   ;; Simple multi-valued data
+   :AncillaryKeywords :AncillaryKeywords
+   :ContactGroups #(mapcat contact-group->keywords (:ContactGroups %))
+   :ContactPersons #(mapcat contact-person->keywords (:ContactPersons %))
+   :Platforms #(mapcat platform->keywords (:Platforms %))
+   :ScienceKeywords #(mapcat science-keyword->keywords (:ScienceKeywords %))
+   :ServiceKeywords #(mapcat service-keyword->keywords (:ServiceKeywords %))
+   :ServiceOrganizations #(mapcat service-organization->keywords (:ServiceOrganizations %))})
+
+(defn flatten-collections
```

Comment:
Could you add a docstring for this? The function name doesn't make a lot of sense to me - in one case you are flattening, in the other case you are creating a collection.

Reply:
Yeah, it's  a bit confusing, since these functions are dealing with:
* single strings extracted from a field
* a collection of strings extracted from a field
* a nested collection of strings extracted from a field and/or its sub-fields
* all of these, but for multiple fields, too

I've added a docstring to clarify the function's purpose and the reason for the conditionally different logic.